internal/runtime/internal/dag/ops.go

```diff
@@ -11,7 +11,9 @@ import (
 // as many edges as possible while maintaining the same "reachability" as the
 // original graph: any node N reachable from node S will still be reachable
 // after a reduction.
-func Reduce(g *Graph) {
+// edgesToPreserve allows the caller to specify a set of edges that
+// must not be removed during the transitive reduction.
+func Reduce(g *Graph, edgesToPreserve map[Edge]struct{}) {
```

Comment:
nit: the reduce function is currently not covered by tests, if you have time it would be nice to add one test where the edge is removed and another test with the same scenario but the edge is added to the edgesToPreserve arg

Reply:
no need for the tests if we decide to get rid of the reduction